src/libraries/Common/src/System/Net/IPv6AddressHelper.Common.cs

```diff
@@ -272,151 +309,155 @@ internal static unsafe bool IsValidStrict(char* name, int start, ref int end)
         //  <member>    numbers
         //      Array filled in with the numbers in the IPv6 groups
         //
-        //  <member>    PrefixLength
-        //      Set to the number after the prefix separator (/) if found
+        //  <member>    scopeId
+        //      Set to the text after the scope separator (%) if found
         //
         // Assumes:
         //  <Name> has been validated and contains only hex digits in groups of
-        //  16-bit numbers, the characters ':' and '/', and a possible IPv4
+        //  16-bit numbers, the characters ':', '/' and '%', and a possible IPv4
         //  address
         //
         // Throws:
         //  Nothing
         //
 
-        internal static void Parse(ReadOnlySpan<char> address, Span<ushort> numbers, int start, ref string? scopeId)
+        internal static void Parse<TChar>(ReadOnlySpan<TChar> address, scoped Span<ushort> numbers, out ReadOnlySpan<TChar> scopeId)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
-            int number = 0;
+            ushort number = 0;
             int index = 0;
             int compressorIndex = -1;
             bool numberIsValid = true;
+            // It's possible for an IPv6 address to be short enough that the cost of Contains outweighs the benefit of short-circuiting the parsing logic.
+            bool shortAddress = address.Length <= 16;
+            bool mayContainIPv4Separator = shortAddress || address.Contains(TChar.CreateTruncating('.'));
```

Comment:
Doing changes like this as part of just implementing IUtf8SpanParseable are making it harder to review and understand what's necessary vs what's opportunistic. Can we keep this PR focused on just what's necessary for the functional additions?

Reply:
I'm happy to do this, but without these changes we see a 10-20ns / 30-50% execution time regression, which I was hoping to avoid. I've made these optimisations in this PR to bring its performance back into line with .NET 8.0.

I can pull these out into a separate PR for separate review, (and we can push the API change into .NET 10.0) or I can revert them and accept the performance regression.

Reply:
> but without these changes we see a 10-20ns / 30-50% execution time regression

Do we understand why?

Reply:
My original assumption was that the overhead was just inherent to the original changes needed. `IPv6AddressHelper` used a switch block for parsing, but a case can only refer to a constant value (which rules out `TChar.ParseTruncating(':')` and similar) so I felt I had to convert it into a set of if/else-if blocks. I assumed that the combination of swapping that structure out and anything required for the method to support generics added a bit of overhead.

This is incorrect. I've revisited the approach to the switch block and re-ran the same benchmark on another machine. Results are broadly similar between .NET 8.0 and my PR, so I'm pretty sure I ran into a measurement error and started trying to optimise based on those figures.

| Method      | Job      | Runtime  | s                    | Mean      | Error    | StdDev   | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|------------ |--------- |--------- |--------------------- |----------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| **TryParse** | **Baseline** | **.NET 8.0** | **::192.168.0.1**        |  **84.17 ns** | **0.238 ns** | **0.186 ns** |  **1.00** |    **0.00** | **0.0031** |      **80 B** |        **1.00** |
| TryParse | CoreRun  | .NET 9.0 | ::192.168.0.1        |  82.11 ns | 0.239 ns | 0.212 ns |  0.98 |    0.00 | 0.0031 |      80 B |        1.00 |
|             |          |          |                      |           |          |          |       |         |        |           |             |
| **TryParse** | **Baseline** | **.NET 8.0** | **037777777777**         |  **31.96 ns** | **0.200 ns** | **0.177 ns** |  **1.00** |    **0.00** | **0.0015** |      **40 B** |        **1.00** |
| TryParse | CoreRun  | .NET 9.0 | 037777777777         |  28.52 ns | 0.123 ns | 0.115 ns |  0.89 |    0.01 | 0.0015 |      40 B |        1.00 |
|             |          |          |                      |           |          |          |       |         |        |           |             |
| **TryParse** | **Baseline** | **.NET 8.0** | **0xff.0x7f.0x20.0x01**  |  **34.44 ns** | **0.175 ns** | **0.136 ns** |  **1.00** |    **0.00** | **0.0015** |      **40 B** |        **1.00** |
| TryParse | CoreRun  | .NET 9.0 | 0xff.0x7f.0x20.0x01  |  34.97 ns | 0.173 ns | 0.154 ns |  1.02 |    0.01 | 0.0015 |      40 B |        1.00 |
|             |          |          |                      |           |          |          |       |         |        |           |             |
| **TryParse** | **Baseline** | **.NET 8.0** | **1:2:3:4:5:6:7:8::**    |  **46.59 ns** | **0.095 ns** | **0.089 ns** |  **1.00** |    **0.00** |      **-** |         **-** |          **NA** |
| TryParse | CoreRun  | .NET 9.0 | 1:2:3:4:5:6:7:8::    |  46.76 ns | 0.039 ns | 0.034 ns |  1.00 |    0.00 |      - |         - |          NA |
|             |          |          |                      |           |          |          |       |         |        |           |             |
| **TryParse** | **Baseline** | **.NET 8.0** | **100:0(...):abcd [22]** | **116.55 ns** | **0.461 ns** | **0.408 ns** |  **1.00** |    **0.00** | **0.0031** |      **80 B** |        **1.00** |
| TryParse | CoreRun  | .NET 9.0 | 100:0(...):abcd [22] | 118.77 ns | 0.576 ns | 0.539 ns |  1.02 |    0.01 | 0.0031 |      80 B |        1.00 |
|             |          |          |                      |           |          |          |       |         |        |           |             |
| **TryParse** | **Baseline** | **.NET 8.0** | **192.168.0.0/16**       |  **24.80 ns** | **0.180 ns** | **0.169 ns** |  **1.00** |    **0.00** |      **-** |         **-** |          **NA** |
| TryParse | CoreRun  | .NET 9.0 | 192.168.0.0/16       |  24.78 ns | 0.272 ns | 0.255 ns |  1.00 |    0.01 |      - |         - |          NA |
|             |          |          |                      |           |          |          |       |         |        |           |             |
| **TryParse** | **Baseline** | **.NET 8.0** | **192.168.0.1**          |  **29.51 ns** | **0.493 ns** | **0.461 ns** |  **1.00** |    **0.00** | **0.0015** |      **40 B** |        **1.00** |
| TryParse | CoreRun  | .NET 9.0 | 192.168.0.1          |  30.50 ns | 0.094 ns | 0.074 ns |  1.03 |    0.02 | 0.0015 |      40 B |        1.00 |
|             |          |          |                      |           |          |          |       |         |        |           |             |
| **TryParse** | **Baseline** | **.NET 8.0** | **4294967295**           |  **24.61 ns** | **0.069 ns** | **0.064 ns** |  **1.00** |    **0.00** | **0.0016** |      **40 B** |        **1.00** |
| TryParse | CoreRun  | .NET 9.0 | 4294967295           |  26.85 ns | 0.118 ns | 0.098 ns |  1.09 |    0.01 | 0.0016 |      40 B |        1.00 |
|             |          |          |                      |           |          |          |       |         |        |           |             |
| **TryParse** | **Baseline** | **.NET 8.0** | **Fe08::1%13542**        |  **79.42 ns** | **0.330 ns** | **0.309 ns** |  **1.00** |    **0.00** | **0.0048** |     **120 B** |        **1.00** |
| TryParse | CoreRun  | .NET 9.0 | Fe08::1%13542        |  69.62 ns | 0.213 ns | 0.189 ns |  0.88 |    0.00 | 0.0031 |      80 B |        0.67 |
|             |          |          |                      |           |          |          |       |         |        |           |             |
| **UriHost**     | **Baseline** | **.NET 8.0** | **http:(...)3/foo [39]** | **413.39 ns** | **0.352 ns** | **0.329 ns** |  **1.00** |    **0.00** | **0.0076** |     **192 B** |        **1.00** |
| UriHost     | CoreRun  | .NET 9.0 | http:(...)3/foo [39] | 412.96 ns | 0.374 ns | 0.350 ns |  1.00 |    0.00 | 0.0076 |     192 B |        1.00 |
|             |          |          |                      |           |          |          |       |         |        |           |             |
| **UriHost**     | **Baseline** | **.NET 8.0** | **http:(...)3/foo [26]** | **253.23 ns** | **0.646 ns** | **0.604 ns** |  **1.00** |    **0.00** | **0.0072** |     **184 B** |        **1.00** |
| UriHost     | CoreRun  | .NET 9.0 | http:(...)3/foo [26] | 243.85 ns | 1.031 ns | 0.965 ns |  0.96 |    0.00 | 0.0072 |     184 B |        1.00 |